perception/sub8_pointcloud/src/OGridGen.cpp

```diff
@@ -0,0 +1,147 @@
+#include "OGridGen.hpp"
+
+//TODO: Add service call to clear ogrid
+
+OGridGen::OGridGen() : nh_(ros::this_node::getName()), classification_(&nh_)
+{
+  pub_grid_ = nh_.advertise<nav_msgs::OccupancyGrid>("/ogrid", 10, true);
+  pub_point_cloud_ = nh_.advertise<sensor_msgs::PointCloud2>("/point_cloud", 1);
+
+  nh_.param<float>("resolution_", resolution_, 0.2f);
+  nh_.param<float>("ogrid_size_", ogrid_size_, 50.f);
+  nh_.param<float>("pool_depth_", pool_depth_, 7.f);
+  
+  int point_cloud_buffer_Size;
+  nh_.param<int>("buffer_size", point_cloud_buffer_Size, 5000);
+  point_cloud_buffer_.set_capacity(point_cloud_buffer_Size);
+
+  service_get_bounds_ = nh_.serviceClient<sub8_msgs::Bounds>("get_bounds_");
+
+  timer_ = nh_.createTimer(ros::Duration(0.3), std::bind(&OGridGen::publish_ogrid, this, std::placeholders::_1));
+  sub_to_imaging_sonar_ = nh_.subscribe("/blueview_driver/ranges", 1, &OGridGen::callback, this);
+
+}
+/*
+  Looped based on timer_.
+  Reads point_cloud_buffer_ and publishes a PointCloud2
+  Reads mat_ogrid_ and publish a OccupencyGrid
+*/
+void OGridGen::publish_ogrid(const ros::TimerEvent &)
+{
+  // Service call for 'get_bounds'
+  sub8_msgs::Bounds get_bound_data;
+  if (service_get_bounds_.call(get_bound_data))
+  {
+    bounds_ = std::vector<cv::Point>();
+    bounds_.reserve(4);
+    // Populate bounds_ with the data from service call
+    for (auto &p : get_bound_data.response.bounds)
+    {
+      bounds_.push_back(cv::Point(p.x / resolution_ + mat_ogrid_.cols / 2, p.y / resolution_ + mat_ogrid_.rows / 2));
+    }
+
+    // Convert bounds_ vector to an array of array and use openCV function to draw a polygon
+    const cv::Point *pts = (const cv::Point *)cv::Mat(bounds_).data;
+    int npts = cv::Mat(bounds_).rows;
+    cv::polylines(mat_ogrid_, &pts, &npts, 1, true, 255, 3, CV_8U, 0);
+  }
+
+  // Flatten the mat_ogrid_ into a 1D vector for OccupencyGrid message
+  nav_msgs::OccupancyGrid rosGrid;
```

Comment:
because this might get big, it would make sense to have this as a class variable that you update. You don't want to be creating another huge vector every time this runs.